Initial attempt at documenting new ODE interface (design-doc #19) (#213)
stan-dev/docs: src/functions-reference/higher-order_functions.Rmd

## Ordinary Differential Equation (ODE) Solvers {#functions-ode-solver}

Stan provides several higher order functions for solving initial value problems specified as Ordinary Differential Equations (ODEs).

> Solving an initial value ODE means given a set of differential equations $y' = f(y, t)$ and initial conditions $y(t_0)$, solving for $y$ at a sequence of times $t_0 < t_1 \leq t_2, \cdots \leq t_n$. $f(y, t)$ is referred to here as the ODE system function.

Reviewer: Here we should make the parameters $\theta$ explicit in f and y, I think... or say that we suppress them on purpose.

\index{{\tt \bfseries ode\_bdf\_tol }!{\tt (function ode, vector initial\_state, real initial\_time, real[] times, data real rel\_tol, data real abs\_tol, data int max\_num\_steps, ...): vector[]}|hyperpage}

`vector[]` **`ode_bdf_tol`**`(function ode, vector initial_state, real initial_time, real[] times, data real rel_tol, data real abs_tol, data int max_num_steps, ...)`<br>\newline
Solves the ODE system for the times provided using the backward
differentiation formula (BDF) method with the implementation from
CVODES with additional control parameters for the CVODES solver.

### Arguments to the ODE solvers

The arguments to the ODE solvers in both the stiff and non-stiff cases
are as follows.

*   *`ode`*: ODE system function with signature `(real, vector, ...):vector`

The arguments represent (1) time, (2) system state, and a sequence of arguments of any type passed unmodified from the integration function call. The return value contains the derivatives with respect to time of the state.

*   *`initial_state`*: initial state, type `vector`,

> *   *`initial_time`*: initial time, type `int` or `real`,

Reviewer: We allow "int" as initial time... really? Is this cast into a real internally?

Reviewer 2: Any function that can take a `real` argument should be able to take an `int` argument by casting it to `real`.  This does not (yet) apply to `real[]` vs. `int[]` arrays, I'm afraid, but should hold for primitives.

Author: This is one of those things that's for convenience as much as anything. Something like:

```
ode_bdf(f, y0, 0, ts, ...)
```

will cause an error if only real is supported because the 0 gets interpreted as an int instead of a double. int should promote to var or double, but I vaguely remember having this problem at one point anyway. Not sure the exact conditions.

*   *`times`*: solution times, type `real[]`,

*   *`...`*: arguments that will be passed through unmodified to the ODE system function. The types here must match the types in the `...` portion of the ode system function arguments.

For the versions of the ode integrator functions ending in `_tol`, these three
parameters must also be provided after `times` and before the `...` arguments:

*   `data`   *`rel_tol`*: relative tolerance for the ODE  solver, type `real`,
data only,

*   `data`   *`abs_tol`*: absolute tolerance for the ODE  solver, type `real`,
data only, and

*   `data`   *`max_num_steps`*: maximum number of steps to take between output
times in the ODE solver, type `int`, data only.

### Specifying the ODE System Function

> The ODE system function must have the a `vector` return type, and the first two arguments must be a `real` and `vector` in that order. These two arguments can be followed by any number of other arguments of any type:

Reviewer: "The ODE system function must have as return type a" ....

```
 vector ode(real time, vector state, ...)
```

`...` are any number of arguments passed through unmodified from the ODE integrate
call.

The ODE system function should return the derivative of the state with
respect to time at the time and state provided. The length of the returned
vector must match the length of the state input into the function.

The arguments to this function are:

*   *`time`*, the time to evaluate the ODE system

*   *`state`*, the state of the ODE system at the time specified

> *   *...*, all other arguments passed through from the integrate call. Each

Reviewer: *`...`*

argument can be any type.

#### Return values

The return value for the ODE solvers is an array of vectors (type `vector[]`),
with values consisting of solutions at the specified times.

#### Sizes and parallel arrays

The sizes must match, and in particular, the following groups are of
the same size:

*   state variables passed into the system function, derivatives
returned by the system function, initial state passed into the
solver, and length of each vector in the output,

*   solution times and number of vectors in the output,

> ## Old integrate_ode_X ODE Interface {#old-functions-ode-solver}

Reviewer: Maybe call this section "Deprecated `integrate_ode_*` ODE Interface" ?

Reviewer 2: It's probably best to start an appendix with deprecated functions as we do in the reference manual.

## Ordinary Differential Equation (ODE) Solvers {#functions-ode-solver}

> Stan provides built-in ordinary differential equation (ODE) solvers.
> Because the first argument to each of the solvers is a function itself,
> the ODE solvers in Stan are considered high order functions.

Reviewer 3: ```suggestion
Stan provides several higher order functions for solving Ordinary Differential Equations (ODEs).
```

### Specifying an Ordinary Differential Equation as a Function

> A system of ODEs is specified as an ordinary function in Stan within
> the functions block. The ODE system function must have the same return
> type and first two arguments as the following function:

Reviewer 3: ```suggestion
In Stan, a system of ODEs is specified as a function. The ODE system function must have the a `vector` return type with the first two arguments to the function being a `real` and `vector` such as:
```

```
 vector ode(real time, vector state, ...)
```

> `...` are any number of arguments passed through unmodified from the ODE integrate
> call.
> 
> The ODE system function should return the derivative of the state with
> respect to time at the time and state provided. The length of the returned
> vector must match the length of the state input into the function.
> 
> The arguments to this function are:
> 
> *   *`time`*, the time to evaluate the ODE system
> 
> *   *`state`*, the state of the ODE system at the time specified
> 
> *   *...*, all other arguments passed through from the integrate call. Each
> argument can be any type.

Reviewer 3: Before this section I would show an example signature for the ODE integrators because

> all other arguments passed through from the integrate call

Doesn't really make sense without knowing what the integrators look like

```suggestion

Where the arguments to the function are:

*   *`time`*, the time to evaluate the ODE system

*   *`state`*, the state of the ODE system at the time specified

*   *...*, Any additional arguments needed in the ODE. Each of these can be arguments of any type.

The ODE system function should return the derivative of the `state` with
respect to `time` at the `time` and `state` provided. The length of the returned
vector must match the length of the state input into the function.
```

Author: I moved the ODE system function definition below the signatures and the arguments to the integrator functions.

It's still slightly confusing because the first argument to the integrator function is the ODE system function, but there's a little inline doc there that hopefully works.

### Non-Stiff Solver

<!-- vector[]; ode_rk45; (function ode, vector initial_state, real initial_time, real[] times, ...); -->
\index{{\tt \bfseries ode\_rk45 }!{\tt (function ode, real[] initial\_state, real initial\_time, real[] times, ...): vector[]}|hyperpage}

`vector[]` **`ode_rk45`**`(function ode, vector initial_state, real initial_time, real[] times, ...)`<br>\newline
Solves the ODE system for the times provided using the Runge Kutta

> Dopri algorithm with the implementation from Boost.

Reviewer 3: Should there be a link to wiki or boost for this method? Or is the user expected to know this solver already?

Author: I made the naming a little clearer, but yeah this is a searchable standard algorithm.

`vector[]` **`ode_bdf`**`(function ode, vector initial_state, real initial_time, real[] times, ...)`<br>\newline
Solves the ODE system for the times provided using the backward
differentiation formula (BDF) method with the implementation from
CVODES.

<!-- vector[]; ode_bdf_tol; (function ode, vector initial_state, real initial_time, real[] times, data real rel_tol, data real abs_tol, data int max_num_steps, ...); -->
\index{{\tt \bfseries ode\_bdf\_tol }!{\tt (function ode, vector initial\_state, real initial\_time, real[] times, data real rel\_tol, data real abs\_tol, data int max\_num\_steps, ...): vector[]}|hyperpage}

`vector[]` **`ode_bdf_tol`**`(function ode, vector initial_state, real initial_time, real[] times, data real rel_tol, data real abs_tol, data int max_num_steps, ...)`<br>\newline
Solves the ODE system for the times provided using the backward
differentiation formula (BDF) method with the implementation from
CVODES with additional control parameters for the CVODES solver.

### Arguments to the ODE solvers

The arguments to the ODE solvers in both the stiff and non-stiff cases
are as follows.

*   *`ode`*: function literal referring to a function specifying  the

> system of differential equations with signature:

Reviewer 3: What does function literal mean? Should we use that same term when talking about the ODE function above?

Reviewer 3: Is this section the same as the one above?

Author: > Is this section the same as the one above?

I left a copy of the old integrator docs there. The function signature will be deprecated but I figure it should still have docs so people know how to convert old models. So there is duplicate information.
